File: sprunk/definitions.py
import os.path

import strictyaml

__all__ = [
    'load_definitions',
]

DEF_KEYS = [
    'name',
    'prefix',
    'include',
    'id',
    'solo',
    'to-ad',
    'to-news',
    'time-morning',
    'time-evening',
    'general',
    'intro',
    'ad',
    'news',
    'music',
]

SPECIAL_KEYS = ['name', 'prefix', 'include', 'intro', 'music']

MUSIC_KEYS = [
    'path',
    'title',
    'artist',
    'album',
    'pre',
    'post',
]

MUSIC_OPTIONAL_KEYS = [
    'album',
]

def load_definitions(files, extension):
    whole = {'music': [], 'name': 'Sprunk', 'intro': []}
    for k in DEF_KEYS:
        if k in SPECIAL_KEYS:
            continue
        whole[k] = list()

    def locate_file(base, fname, ext=None):
        if ext is None:
            ext = '.' + extension
        return os.path.abspath(os.path.join(base, fname) + ext)

    def locate_files(base, data, key):
        l = data.get(key, [])
        for n in l:
            path = locate_file(base, n)
            if not path in whole[key]:
                whole[key].append(path)

    def parse_timestamp(s):
        a, b = s.split(':', 1)
        return int(a, base=10) * 60 + float(b)

    def handle_file(fname):
        with open(fname) as f:
            data = strictyaml.load(f.read()).data
        base = os.path.split(fname)[0]
        prefix = data.get('prefix')
        if prefix:
            base = os.path.join(base, prefix)

        for sub_fname in data.get('include', []):
            full_fname = locate_file(base, sub_fname, ext='')
            handle_file(full_fname)

        if 'name' in data:
            whole['name'] = data['name']

        for k in data:
            if k not in DEF_KEYS:
                raise RuntimeError('unknown key {} in file {}'.format(k, fname))
        for k in DEF_KEYS:
            if k in SPECIAL_KEYS:
                continue
            locate_files(base, data, k)
        for m in data.get('music', []):
            for k in m:
                if k not in MUSIC_KEYS:
                    raise RuntimeError('unknown key {} in file {}'.format(k, fname))
            for k in MUSIC_KEYS:
                if not k in m and k not in MUSIC_OPTIONAL_KEYS:
                    raise RuntimeError('missing key {} in file {}'.format(k, fname))
            m['path'] = locate_file(base, m['path'])
            for k in MUSIC_OPTIONAL_KEYS:
                if not k in m:
                    m[k] = None
            m['pre'] = parse_timestamp(m['pre'])
            m['post'] = parse_timestamp(m['post'])
            if not [n for n in whole['music'] if n['path'] == m['path']]:
                whole['music'].append(m)
        for i in data.get('intro', []):
            for k in i:
                if k not in MUSIC_KEYS:
                    raise RuntimeError('unknown key {} in file {}'.format(k, fname))
            if 'path' not in i:
                raise RuntimeError('missing key {} in file {}'.format('path', fname))

            i['path'] = locate_file(base, i['path'])
            if not [n for n in whole['intro'] if n['path'] == i['path']]:
                whole['intro'].append(i)

    for fname in files:
        handle_file(fname)

    # resolve intros
    def lookup_song(meta):
        found = None
        for m in whole['music']:
            for k in meta:
                if k == 'path':
                    continue
                if meta[k] != m[k]:
                    break
            else:
                # found a match!
                if found is not None:
                    raise RuntimeError('ambiguous intro match for {}'.format(meta))
                found = m
        if found is None:
            raise RuntimeError('no music matches intro for {}'.format(meta))
        return found
    for i in whole['intro']:
        m = lookup_song(i)
        if not 'intro' in m:
            m['intro'] = list()
        m['intro'].append(i['path'])

    return whole

def lint(defs):
    def check_all(fs):
        for f in fs:
            check(f)

    def check(f):
        if not (os.path.exists(f) and os.path.isfile(f)):
            print('NOT FOUND: {}'.format(f))
            return 1

    for k in DEF_KEYS:
        if k in SPECIAL_KEYS:
            continue
        check_all(defs[k])

    for m in defs['music']:
        check(m['path'])
        for i in m.get('intro', []):
            check(i)
    for i in defs['intro']:
        check(i['path'])

    print('ok!')

    return 0

        

File: src/definitions.rs
use std::path::{Path, PathBuf};

use strict_yaml_rust::{StrictYaml, StrictYamlLoader};

use crate::normalize::normalize;

#[derive(Debug, Clone)]
pub struct Definitions {
    pub paths: Vec<PathBuf>,
    pub name: Option<String>,
    pub solo: Vec<PathBuf>,
    pub general: Vec<PathBuf>,
    pub to_ad: Vec<PathBuf>,
    pub to_news: Vec<PathBuf>,
    pub time_evening: Vec<PathBuf>,
    pub time_morning: Vec<PathBuf>,
    pub id: Vec<PathBuf>,
    pub ad: Vec<PathBuf>,
    pub news: Vec<PathBuf>,
    pub intro: Vec<Intro>,
    pub music: Vec<Song>,
}

#[derive(Debug, Clone)]
pub struct Metadata {
    pub title: String,
    pub artist: String,
    pub album: Option<String>,
}

#[derive(Debug, Clone)]
pub struct Intro {
    pub path: PathBuf,
    pub metadata: Metadata,
}

#[derive(Debug, Clone)]
pub struct Song {
    pub path: PathBuf,
    pub metadata: Metadata,
    pub pre: f32,
    pub post: f32,
}

impl Definitions {
    pub fn open<PI, P>(paths: PI) -> anyhow::Result<Self>
    where
        PI: Iterator<Item = P>,
        P: AsRef<Path>,
    {
        let mut defs = Definitions::empty();
        defs.paths = paths.map(|p| p.as_ref().to_owned()).collect();
        defs.reload()?;
        Ok(defs)
    }

    pub fn empty() -> Self {
        Definitions {
            paths: vec![],
            name: None,
            solo: vec![],
            general: vec![],
            to_ad: vec![],
            to_news: vec![],
            time_evening: vec![],
            time_morning: vec![],
            id: vec![],
            ad: vec![],
            news: vec![],
            intro: vec![],
            music: vec![],
        }
    }

    pub fn reload(&mut self) -> anyhow::Result<()> {
        let mut new = Definitions::empty();
        for path in self.paths.iter() {
            new.merge(Definitions::load_one(path)?);
        }
        std::mem::swap(&mut new.paths, &mut self.paths);
        *self = new;
        Ok(())
    }

    fn load_one(path: &PathBuf) -> anyhow::Result<Self> {
        let mut new = Definitions::empty();
        let base = path.parent().unwrap_or(Path::new("."));
        let contents = std::fs::read_to_string(&path)?;
        let datawhole = StrictYamlLoader::load_from_str(&contents)?;
        let data = datawhole
            .get(0)
            .ok_or_else(|| anyhow::anyhow!("could not get definition document"))?;
        Self::check_keys(
            data,
            &[
                "name",
                "prefix",
                "include",
                "solo",
                "general",
                "to-ad",
                "to-news",
                "time-evening",
                "time-morning",
                "id",
                "ad",
                "news",
                "intro",
                "music",
            ],
        )?;

        let prefix = base.join(Self::get_str(data, "prefix")?.unwrap_or("."));

        // read and merge includes first
        if let Some(includes) = Self::get_vec(data, "include")? {
            for inc in includes.iter() {
                let inc = inc
                    .as_str()
                    .ok_or_else(|| anyhow::anyhow!("includes must be strings"))?;
                let inc = normalize(&prefix.join(inc));
                new.merge(Definitions::load_one(&inc)?);
            }
        }

        // read the radio name
        if let Some(name) = Self::get_str(data, "name")? {
            new.name = Some(name.to_owned());
        }

        // read in simple path lists
        new.solo.extend(Self::get_path_vec(data, "solo", &prefix)?);
        new.general
            .extend(Self::get_path_vec(data, "general", &prefix)?);
        new.to_ad
            .extend(Self::get_path_vec(data, "to-ad", &prefix)?);
        new.to_news
            .extend(Self::get_path_vec(data, "to-news", &prefix)?);
        new.time_evening
            .extend(Self::get_path_vec(data, "time-evening", &prefix)?);
        new.time_morning
            .extend(Self::get_path_vec(data, "time-morning", &prefix)?);
        new.id.extend(Self::get_path_vec(data, "id", &prefix)?);
        new.ad.extend(Self::get_path_vec(data, "ad", &prefix)?);
        new.news.extend(Self::get_path_vec(data, "news", &prefix)?);

        // read intros
        if let Some(intros) = Self::get_vec(data, "intro")? {
            new.intro.reserve(intros.len());
            for intro in intros.iter() {
                Self::check_keys(intro, &["path", "title", "artist", "album"])?;
                let path = Self::get_str(intro, "path")?
                    .ok_or_else(|| anyhow::anyhow!("song requires path"))?;
                let path = Self::verify_media(&prefix.join(path))?;
                let title = Self::get_str(intro, "title")?
                    .ok_or_else(|| anyhow::anyhow!("song requires title"))?
                    .to_owned();
                let artist = Self::get_str(intro, "artist")?
                    .ok_or_else(|| anyhow::anyhow!("song requires artist"))?
                    .to_owned();
                let album = Self::get_str(intro, "album")?.map(|s| s.to_owned());
                new.intro.push(Intro {
                    path,
                    metadata: Metadata {
                        title,
                        artist,
                        album,
                    },
                })
            }
        }

        // read music
        if let Some(songs) = Self::get_vec(data, "music")? {
            new.music.reserve(songs.len());
            for song in songs.iter() {
                Self::check_keys(song, &["path", "title", "artist", "album", "pre", "post"])?;
                let path = Self::get_str(song, "path")?
                    .ok_or_else(|| anyhow::anyhow!("song requires path"))?;
                let path = Self::verify_media(&prefix.join(path))?;
                let title = Self::get_str(song, "title")?
                    .ok_or_else(|| anyhow::anyhow!("song requires title"))?
                    .to_owned();
                let artist = Self::get_str(song, "artist")?
                    .ok_or_else(|| anyhow::anyhow!("song requires artist"))?
                    .to_owned();
                let album = Self::get_str(song, "album")?.map(|s| s.to_owned());
                let pre = Self::get_str(song, "pre")?
                    .ok_or_else(|| anyhow::anyhow!("song requires pre"))?;
                let post = Self::get_str(song, "post")?
                    .ok_or_else(|| anyhow::anyhow!("song requires post"))?;
                new.music.push(Song {
                    path,
                    metadata: Metadata {
                        title,
                        artist,
                        album,
                    },
                    pre: Self::parse_time(pre)?,
                    post: Self::parse_time(post)?,
                })
            }
        }

        Ok(new)
    }

    pub fn merge(&mut self, other: Definitions) {
        if other.name.is_some() {
            self.name = other.name;
        }
        self.solo.extend(other.solo);
        self.general.extend(other.general);
        self.to_ad.extend(other.to_ad);
        self.to_news.extend(other.to_news);
        self.time_evening.extend(other.time_evening);
        self.time_morning.extend(other.time_morning);
        self.id.extend(other.id);
        self.ad.extend(other.ad);
        self.news.extend(other.news);
        self.intro.extend(other.intro);
        self.music.extend(other.music);
    }

    pub fn verify(&self) -> anyhow::Result<()> {
        // make sure each intro matches a song
        'intro: for intro in self.intro.iter() {
            for song in self.music.iter() {
                if Self::meta_match(&intro.metadata, &song.metadata) {
                    continue 'intro;
                }
            }
            anyhow::bail!("intro does not match any song: {:?}", intro.path);
        }
        Ok(())
    }

    fn meta_match(a: &Metadata, b: &Metadata) -> bool {
        if let Some(ref aa) = a.album {
            if let Some(ref ab) = b.album {
                if aa != ab {
                    return false;
                }
            }
        }
        a.artist == b.artist && a.title == b.title
    }

    pub fn get_intros<'a>(&'a self, meta: &'a Metadata) -> impl Iterator<Item = &'a Intro> {
        self.intro
            .iter()
            .filter(move |i| Self::meta_match(&i.metadata, meta))
    }

    fn get_str<'a>(data: &'a StrictYaml, k: &str) -> anyhow::Result<Option<&'a str>> {
        let v = if data[k].is_badvalue() {
            Some(None)
        } else {
            data[k].as_str().map(Some)
        };
        v.ok_or_else(|| anyhow::anyhow!("bad value for {:?}, expected string", k))
    }

    fn get_vec<'a>(data: &'a StrictYaml, k: &str) -> anyhow::Result<Option<&'a Vec<StrictYaml>>> {
        let v = if data[k].is_badvalue() {
            Some(None)
        } else {
            data[k].as_vec().map(Some)
        };
        v.ok_or_else(|| anyhow::anyhow!("bad value for {:?}, expected list", k))
    }

    fn get_path_vec(data: &StrictYaml, k: &str, prefix: &Path) -> anyhow::Result<Vec<PathBuf>> {
        if let Some(paths) = Self::get_vec(data, k)? {
            let mut ret = Vec::with_capacity(paths.len());
            for p in paths.iter() {
                let p = p
                    .as_str()
                    .ok_or_else(|| anyhow::anyhow!("bad path in {:?}, expected string", k))?;
                ret.push(Self::verify_media(&prefix.join(Path::new(p)))?);
            }
            Ok(ret)
        } else {
            Ok(vec![])
        }
    }

    fn verify_media(path: &Path) -> anyhow::Result<PathBuf> {
        let mut mutpath = normalize(path);
        let exts = &["flac", "wav", "ogg"];
        for ext in exts {
            mutpath.set_extension(ext);
            if mutpath.exists() {
                return Ok(mutpath);
            }
        }
        anyhow::bail!("file does not exist: {:?} (tried: {:?})", path, exts);
    }

    fn parse_time(time: &str) -> anyhow::Result<f32> {
        let mut r = 0.0;
        for part in time.split(":") {
            r *= 60.0;
            r += part
                .parse::<f32>()
                .map_err(|_| anyhow::anyhow!("bad timestamp: {:?}", time))?;
        }
        Ok(r)
    }

    fn check_keys(data: &StrictYaml, keys: &[&str]) -> anyhow::Result<()> {
        let hash = data
            .as_hash()
            .ok_or_else(|| anyhow::anyhow!("expected yaml dictionary"))?;
        for k in hash.keys() {
            if let Some(k) = k.as_str() {
                if !keys.contains(&k) {
                    anyhow::bail!("unknown key {:?}", k);
                }
            } else {
                anyhow::bail!("unknown key {:?}", k);
            }
        }
        Ok(())
    }
}
